toolchain/parse/handle_decl_scope_loop.cpp

```diff
@@ -35,71 +77,182 @@ static auto HandleUnrecognizedDecl(Context& context, int32_t subtree_start)
+    return true;
+  } else if (!TokenIsModifierOrIntroducer(
+                 context.PositionKind(Lookahead::NextToken))) {
+    // TODO: If the next token isn't a colon or `class`, try to recover
+    // based on whether we're in a class, whether we have an `extend`
+    // modifier, and the following tokens.
+    context.AddLeafNode(NodeKind::InvalidParse, context.Consume(),
+                        /*has_error=*/true);
+    CARBON_DIAGNOSTIC(ExpectedAfterBase, Error,
+                      "`class` or `:` expected after `base`.");
+    context.emitter().Emit(*context.position(), ExpectedAfterBase);
+    OutputInvalidParseSubtree(context, state.subtree_start);
+    return true;
+  } else {
+    // `base` is considered a declaration modifier if it is followed by
+    // another modifier or an introducer.
+    return false;
+  }
+}
 
-  switch (context.PositionKind()) {
-    case Lex::TokenKind::CloseCurlyBrace:
-    case Lex::TokenKind::FileEnd: {
-      // This is the end of the scope, so the loop state ends.
-      context.PopAndDiscardState();
-      return;
+// Returns true if the current position is a declaration, handling it if so.
+static auto TryHandleAsDecl(Context& context, Context::StateStackEntry state,
+                            bool saw_modifier, Lex::TokenKind position_kind)
+    -> bool {
+  auto introducer = [&](NodeKind node_kind, State next_state) {
+    context.ReplacePlaceholderNode(state.subtree_start, node_kind,
+                                   context.Consume());
+    // Reuse state here to retain its `subtree_start`.
+    state.state = next_state;
+    context.PushState(state);
+  };
+
+  switch (position_kind) {
+    case Lex::TokenKind::Base: {
+      return TryHandleBaseAsDecl(context, state);
     }
-    // `import`, `library`, and `package` manage their packaging state.
-    case Lex::TokenKind::Import: {
-      context.PushState(State::Import);
-      return;
+
+    case Lex::TokenKind::Extend: {
+      if (TokenIsModifierOrIntroducer(
+              context.PositionKind(Lookahead::NextToken))) {
+        // `extend` is considered a declaration modifier if it is followed by
+        // another modifier or an introducer.
+        return false;
+      } else {
+        // TODO: Treat this `extend` token as a declaration introducer
+        HandleUnrecognizedDecl(context, state.subtree_start);
+        return true;
+      }
     }
-    case Lex::TokenKind::Library: {
-      context.PushState(State::Library);
-      return;
+
+    case Lex::TokenKind::Impl: {
+      if (TokenIsModifierOrIntroducer(
+              context.PositionKind(Lookahead::NextToken))) {
+        // `impl` is considered a declaration modifier if it is followed by
+        // another modifier or an introducer.
+        return false;
+      } else {
+        // TODO: Treat this `impl` token as a declaration introducer
+        HandleUnrecognizedDecl(context, state.subtree_start);
+        return true;
+      }
     }
```

Comment:
It seems like there's still duplication here; can we check for the ambiguous cases first somehow? In all three cases, if a token is both an introducer and a modifier, the logic is the same: it's a modifier if the next token is a modifier or an introducer.